catroid/src/main/java/org/catrobat/catroid/io/StorageOperations.java

```diff
@@ -0,0 +1,354 @@
+		cursor.moveToFirst();
+		int index = cursor.getColumnIndex(MediaStore.MediaColumns.DATA);
+		try {
+			path = cursor.getString(index);
+		} catch (CursorIndexOutOfBoundsException e) {
+			Log.e(TAG, Log.getStackTraceString(e));
+		} finally {
+			cursor.close();
+		}
+		return path;
+	}
+
+	public static String getSanitizedFileName(File file) {
+		if (file.isDirectory()) {
+			return file.getName();
+		}
+
+		String name = file.getName();
+		int extensionStartIndex = name.lastIndexOf('.');
+		int appendixStartIndex = name.lastIndexOf(FILE_NAME_APPENDIX);
+
+		if (appendixStartIndex == -1) {
+			appendixStartIndex = extensionStartIndex;
+		}
+
+		if (appendixStartIndex == -1) {
+			return name;
+		}
+
+		return name.substring(0, appendixStartIndex);
+	}
+
+	public static File copyFile(File srcFile) throws IOException {
+		File dstDir = srcFile.getParentFile();
+		return copyFileToDirectory(srcFile, dstDir);
+	}
+
+	public static File copyFileToDirectory(File srcFile, File dstDir) throws IOException {
+		if (!srcFile.exists()) {
+			throw new FileNotFoundException("File: " + srcFile.getAbsolutePath() + " does not exist.");
+		}
+
+		if (!dstDir.exists()) {
+			throw new FileNotFoundException("File: " + dstDir.getAbsolutePath() + " does not exist.");
+		}
+
+		if (!dstDir.isDirectory()) {
+			throw new IOException("Destination: " + dstDir.getAbsolutePath() + " is not a directory.");
+		}
+
+		File dstFile = getUniqueFile(srcFile.getName(), dstDir);
+		transferFileToFile(srcFile, dstFile);
+
+		return dstFile;
+	}
+
+	public static File copyStreamToFile(InputStream inputStream, File dstFile) throws IOException {
+		if (!dstFile.exists() && !dstFile.createNewFile()) {
+			throw new IOException("Could not create dstFile at: " + dstFile.getAbsolutePath() + ".");
+		}
+
+		return transferStreamToFile(inputStream, dstFile);
+	}
+
+	public static File createFileFromStreamInDir(InputStream inputStream, File dstDir, String fileName) throws
+			IOException, InvalidPathException {
+		if (!dstDir.exists()) {
+			throw new FileNotFoundException("Directory: " + dstDir.getAbsolutePath() + " does not exist.");
+		}
+
+		if (!dstDir.isDirectory()) {
+			throw new IOException(dstDir.getAbsolutePath() + " is not a directory.");
+		}
+
+		File dstFile = getUniqueFile(fileName, dstDir);
+
+		if (!dstFile.createNewFile()) {
+			throw new IOException("Could not create dstFile at: " + dstFile.getAbsolutePath() + ".");
+		}
+
+		return transferStreamToFile(inputStream, dstFile);
+	}
+
+	public static void transferFileToFile(File src, File dst) throws IOException {
+		FileChannel ic = new FileInputStream(src).getChannel();
+		FileChannel oc = new FileOutputStream(dst).getChannel();
+
+		try {
+			ic.transferTo(0, ic.size(), oc);
+		} finally {
+			if (ic != null) {
+				ic.close();
+			}
+			oc.close();
+		}
+	}
+
+	private static File transferStreamToFile(InputStream inputStream, File dstFile) throws IOException {
```

Comment:
What do you think about `transferData`, alternatively `transferDataToFile`?
`transferFileToFile` and `transferStreamToFile` are kinda Hungarian notation.

Reply:
Ah, I just realized, this is really just splitting things up by moving code from one class to another. Ignore my comment :D

Reply:
I'm happy about the input, so feel free to comment on other things as well.